packages/yarnpkg-pnp/sources/loader/makeManager.ts

```diff
@@ -96,26 +96,39 @@ export function makeManager(pnpapi: PnpApi, opts: MakeManagerOptions) {
   }
 
   function findApiPathFor(modulePath: NativePath) {
-    const controlledBy: Array<PortablePath> = [];
+    const candidates: Array<{packageLocation: NativePath, apiPath: PortablePath}> = [];
+
     for (const [apiPath, apiEntry] of apiMetadata) {
       const locator = apiEntry.instance.findPackageLocator(modulePath);
 
-      if (locator) {
-        if (apiMetadata.size === 1) {
-          return apiPath;
-        } else {
-          controlledBy.push(apiPath);
-        }
-      }
+      if (!locator)
+        continue;
+
+      if (apiMetadata.size === 1)
+        return apiPath;
+
+      const packageInformation = apiEntry.instance.getPackageInformation(locator);
+      if (!packageInformation)
+        throw new Error(`Assertion failed: Couldn't get package information for '${modulePath}'`);
+
+      candidates.push({packageLocation: packageInformation.packageLocation, apiPath});
     }
 
-    if (controlledBy.length !== 0) {
-      if (controlledBy.length === 1)
-        return controlledBy[0];
+    if (candidates.length === 1)
+      return candidates[0].apiPath;
+
+    if (candidates.length > 1) {
+      const [candidateA, candidateB] = [...candidates].sort(
+        (a, b) => b.packageLocation.length - a.packageLocation.length
+      );
+
+      if (candidateA.packageLocation.length > candidateB.packageLocation.length)
+        return candidateA.apiPath;
 
       throw new Error(
-        `Unable to locate pnpapi, the module '${modulePath}' is controlled by multiple pnpapi instances.\nThis is usually caused by using the global cache (enableGlobalCache: true)\n\nControlled by:\n${controlledBy
-          .map(pnpPath => `  ${npath.fromPortablePath(pnpPath)}`)
+        `Unable to locate pnpapi, the module '${modulePath}' is controlled by multiple pnpapi instances.\n` +
+        `This is usually caused by using the global cache (enableGlobalCache: true)\n\nControlled by:\n${candidates
```

Comment:
I know it's not this PR that introduced this error message (I missed this part before), but I'd prefer `@yarnpkg/pnp` to not issue Yarn-specific error messages - it would be weird for a pnpm user that uses PnP to see things referencing `enableGlobalCache`.

Unfortunately, I can't propose any solution other than having `@yarnpkg/plugin-pnp` transform the hook's source code before writing it to disk (which would add a lot of complexity).